File: Cpp/525_Contiguous_Array.cpp
class Solution {
public:
    int findMaxLength(vector<int>& nums) {
        unordered_map<int, int> start;
        int result = 0;
        int sum = 0;
        start[0] = -1;
        for(int i=0; i<nums.size(); ++i){
            sum += nums[i] == 1 ? 1 : -1; 
            auto found = start.find(sum);
            if (found != start.end()){
                result = std::max(result, i - found->second);
            } else {
                start[sum] = i;
            }
        }
        
        return result;
    }
};

File: Kotlin/525_Contiguous_Array.kt
class Solution {
    fun findMaxLength(nums: IntArray): Int {
        val start = HashMap<Int, Int>()
        var result = 0
        var sum = 0
    
        start[0] = -1
        for(i in nums.indices) {
            sum += if(nums[i] == 0) - 1 else 1

            if(start.contains(sum)) {
                result = maxOf(result, i - start[sum]!!)
            } else {
                start[sum] = i
            }
        }

        return result
    }
}